### ISSUE TITLE
[LinOTP 2.10.5.2] sha1 TOTP reuse with LinOTP

### ISSUE THREAD

**Issue by sfoutrel** [LinOTP/LinOTP#110]
Hello,
I discovered that after a certain amount of time a TOTP is accepted instead of being rejected as already used and timeshift value for token is modified negatively.

I think it's related to the "automatic resync" but I can't find an acceptable combination that does trigger this (bug/feature).
Any help would be appreciated.

Thanks for your work.

**Comment by sfoutrel** [comment]
as a follow up I cannot find a trace of token resyncing in the linotp logs

**Comment by sfoutrel** [comment]
TokenInfo.timeShift 0.0
time,TOTPValue,Result,log
17:35:32,817599,OK,[linotp.lib.auth.finishtokens][finish_valid_tokens #157] user u'sfoutrel'@u'otp.local' successfully authenticated.
17:53:49,817599,KO,[linotp.lib.auth.finishtokens][finish_checked_tokens #114] user u'sfoutrel'@u'otp.local' failed to auth.
17:56:04,817599,OK,[linotp.lib.auth.finishtokens][finish_valid_tokens #157] user u'sfoutrel'@u'otp.local' successfully authenticated.
TokenInfo.timeShift -1230.0
tokeninfo.timewindow 300
linotp.DefaultCountWindow 100
linotp.AutoResync True
linotp.AutoResyncTimeout 240
linotp.DefaultCountWindow 10

**Comment by authprivsec** [comment]
We were able to reproduce and hotfix this error. Thank you for reporting this issue and providing diagnostics.

The fix is in commit [c52586ab7d3ffbbd7a18b325586d0f0a45631d14](https://github.com/LinOTP/LinOTP/commit/c52586ab7d3ffbbd7a18b325586d0f0a45631d14)

There will be a refactored version of the automatic resync function in master, soon.

We published https://www.linotp.org/CVE-2019-12887.txt for this vulnerability and a patch and installation instructions at https://linotp.org/howtos/howto-hotfix-autoresync.html
